compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1157,72 +1178,76 @@ public void visit(BLangResource resourceNode) {
-        }
-        return isExportableType(type);
-    }
+    private void checkForExportableType(BTypeSymbol symbol, DiagnosticPos pos) {
 
-    private boolean isExportableType(BType type) {
-        if (type.tag == TypeTags.OBJECT) {
-            // All fields and methods need to be public in order for a object to qualify to be exportable
-            // (When not specified to be public)
-            for (BField field : ((BObjectType) type).fields) {
-                if (!Symbols.isPublic(field.symbol)) {
-                    return false;
+        if (symbol == null || symbol.type == null || Symbols.isFlagOn(symbol.flags, Flags.TYPE_PARAM)) {
+            // This is a built-in symbol or a type Param.
+            return;
+        }
+        switch (symbol.type.tag) {
+            case TypeTags.ARRAY:
+                checkForExportableType(((BArrayType) symbol.type).eType.tsymbol, pos);
+                return;
+            case TypeTags.TUPLE:
+                BTupleType tupleType = (BTupleType) symbol.type;
+                tupleType.tupleTypes.forEach(t -> checkForExportableType(t.tsymbol, pos));
+                if (tupleType.restType != null) {
+                    checkForExportableType(tupleType.restType.tsymbol, pos);
                 }
-                for (BAttachedFunction attachedFunc : ((BObjectTypeSymbol) type.tsymbol).attachedFuncs) {
-                    if (!Symbols.isPublic(attachedFunc.symbol)) {
-                        return false;
+                return;
+            case TypeTags.MAP:
+                checkForExportableType(((BMapType) symbol.type).constraint.tsymbol, pos);
+                return;
+            case TypeTags.RECORD:
+                if (Symbols.isFlagOn(symbol.flags, Flags.ANONYMOUS)) {
+                    BRecordType recordType = (BRecordType) symbol.type;
+                    recordType.fields.forEach(f -> checkForExportableType(f.type.tsymbol, pos));
+                    if (recordType.restFieldType != null) {
+                        checkForExportableType(recordType.restFieldType.tsymbol, pos);
                     }
+                    return;
                 }
-            }
-        } else if (type.tag == TypeTags.TYPEDESC) {
-            return isExportableType(((BTypedescType) type).constraint);
+                break;
+            case TypeTags.TABLE:
+                BTableType tableType = (BTableType) symbol.type;
+                if (tableType.constraint != null) {
+                    checkForExportableType(tableType.constraint.tsymbol, pos);
+                }
+                return;
+            case TypeTags.STREAM:
+                BStreamType streamType = (BStreamType) symbol.type;
+                if (streamType.constraint != null) {
+                    checkForExportableType(streamType.constraint.tsymbol, pos);
+                }
+                return;
+            // TODO : Add support for other types. such as union and objects
```

Comment:
Created issue to track this. : https://github.com/ballerina-platform/ballerina-lang/issues/18342